### DISCUSSION TITLE
Should we switch to standards-based package builds?

### DISCUSSION THREAD

**Discussion by kandersolar** [pvlib/pvlib-python#1486] [Ideas]
This post is a brain dump of what I've learned about previous, current, and future best practice for python packaging.  Nothing here is at all urgent as far as I'm aware, so feel free to bookmark this for a rainy day; I just wanted to lay it out while it's still fresh.

## Evolution of the python packaging tools ecosystem

When you install a package via `pip install ...`, part of what's going on under the hood involves a distribution in the form of either an `sdist` or a `bdist_wheel`.  We currently generate our `sdist` and `bdist_wheel` files (`.tar.gz` and `.whl` files on [PyPI](https://pypi.org/project/pvlib/#files)) using [setuptools](https://setuptools.pypa.io/en/latest/index.html), the de facto standard in the python packaging community for many years now.  `setuptools`, its particular interface (using either [setup.py](https://github.com/pvlib/pvlib-python/blob/master/setup.py#L109) or [setup.cfg](https://github.com/pvlib/twoaxistracking/blob/main/setup.cfg)), and the underlying `distutils` stdlib all have to maintain backwards compatibility with essentially the entire python ecosystem, making those projects difficult to maintain and extend as the needs of the python packaging community evolve.  It's also difficult to switch away from `setuptools` to some other tool since `pip` only knows how to call `setuptools`.

Or at least that's what was claimed by [PEP 517](https://peps.python.org/pep-0517/), the first in a sequence of Python Enhancement Proposals (PEPs) regarding standardization of a new backend-agnostic build interface for python packaging, where "backend" refers to any tool like `setuptools` that can turn a source tree into `sdist` and `bdist_wheel` distributions suitable to host on PyPI.  I don't necessarily suggest reading them, but the most relevant PEPs to date are [PEP 517](https://peps.python.org/pep-0517/), [PEP 518](https://peps.python.org/pep-0518/), and [PEP 621](https://peps.python.org/pep-0621/).  The idea here is that rather than having `pip` run a build tool-specific command like `python setup.py bdist_wheel` to generate a wheel file, `pip` would instead look up the desired build tool in some static configuration file and activate that tool via a standardized interface.  The configuration file defined in those PEPs is `pyproject.toml`.

Although `setuptools` used to be really the only reasonable option, the acceptance of those PEPs prompted several alternatives to come about, including in no particular order [poetry](https://python-poetry.org/), [flit](https://flit.pypa.io/en/latest/index.html), [enscons](https://github.com/dholth/enscons), [hatch](https://hatch.pypa.io/latest/), [pdm](https://pdm.fming.dev/latest/), and likely others I've not happened to hear about.  I think all of these support the interface defined by those three PEPs (excepting poetry, which doesn't yet support PEP 621, see https://github.com/python-poetry/poetry/issues/3332).  Anyway the point is that `setuptools` is no longer the only practical option for building distributions, and in theory it's easy to switch between the alternatives so long as you use the standardized `pyproject.toml` and `python -m build --sdist --wheel` instead of the `setuptools`-specific `python setup.py sdist bdist_wheel`.

As of https://github.com/pypa/packaging.python.org/pull/1031, the [python packaging guide](https://packaging.python.org/en/latest/tutorials/packaging-projects/) no longer mentions the `setuptools`-specific `setup.py` and `setup.cfg` files, instead sticking entirely to the standards-based interface and configuration described above.  Furthermore, `setuptools` itself is also onboard with the new interface and `pyproject.toml` configuration (https://github.com/pypa/setuptools/pull/3371 and https://github.com/pypa/setuptools/issues/3214).  So it seems like `pyproject.toml` is the future, and likely something we'll want or need to adopt at some point.

## What does this mean for pvlib?

I see three things worth thinking about in light of the above info:

**Should we change how we _invoke_ setuptools**, i.e. do we stop using `python setup.py sdist bdist_wheel` (keeping `setup.py` itself) and switch to `python -m build --sdist --wheel`?  I think probably we should, there's little reason not to, plus it sets a good example for any projects that copy pvlib's setup (which is more than a few!).  [twoaxistracking](https://github.com/pvlib/twoaxistracking/blob/main/.github/workflows/publish.yml#L31) already does this.

**Should we change how we _configure_ setuptools**, i.e. do we switch from `setup.py` to `pyproject.toml`?  I think there's basically no chance that our `setup.py` is going to stop working anytime soon, so there's no real rush to switch.  I do think it's worth looking into switching at some point, but it might not even be possible to switch unless we dump the `spa_c` extension stuff (see [this forum discussion](https://discuss.python.org/t/modern-way-to-build-packages-with-c-cython-extensions/15050)).  Note that `pyproject.toml` is evolving and gaining capabilities over time; if not today, I expect at some point it will be capable of doing what we need. 

**Should we change what backend we're using**, i.e. do we switch from `setuptools` to one of the other backends like `flit`?  I think until we're confident that `pyproject.toml` can do what we need, there's little point in considering alternative backends.  And without a compelling reason to switch to another backend tool, I lean towards "don't fix what ain't broke" here.

**Comment by mdeceglie** [comment]
Last time I used `pyproject.toml` I was disappointed too learn that editable mode installs (`pip install -e`) didn't work. I haven't revisited the question to figure out if there is a solution, but it's something to consider. This is can be a big frustration for development.

**Reply by kandersolar** [reply]
Thanks for bringing that up @mdeceglie.  I don't use editable installs myself so I had forgotten about this point.  I think I remember seeing a workaround with a shim `setup.py`, but either way it seems like setuptools will support editable installs out of the box in the near future: https://discuss.python.org/t/help-testing-pep-660-support-in-setuptools/16904

**Reply by wholmgren** [reply]
> I don't use editable installs myself so I had forgotten about this point. 

Now I am dying to know what is your pvlib development workflow

**Reply by kandersolar** [reply]
https://xkcd.com/763/ and also https://xkcd.com/1172/ :P

**Comment by jules-ch** [comment]
Hello there, I think it's a good idea to switch to PEP517/518 based builds.
Most scientific python packages have switched to this.
It's pretty straight forward, the only bump into this road is the use of versionneer see : https://github.com/python-versioneer/python-versioneer/issues/193

Pandas is commenting the build-backend of `pyproject.toml` for this reason : https://github.com/pandas-dev/pandas/blob/main/pyproject.toml

We can switch to setuptools-scm, matplotlib has made the switch : https://github.com/matplotlib/matplotlib/pull/18971
